Fix previous marks + work with data in class DataManager in config.py

TheMoosnter opened the pull request:
Fix previous marks from first PR + transfer of functions which work with data files in new class DataManager in file config.py

TheMoosnter committed:
fix logs

TheMoosnter committed:
student list in .csv file

TheMoosnter committed:
data load in config file

TheMoosnter committed:
loading of student list in object creating

TheMoosnter committed:
advanced system of logging

hnkovr commented on config.py:
```diff
@@ -1,23 +1,119 @@
+        """
+        return self.data.get(key, default)
+
+    def set(self, key, value):
```
Для следующих доработок или ботов попробуй использовать @property для такого

hnkovr commented on config.py:
```diff
@@ -1,23 +1,119 @@
+            self.data[key].remove(value)
+            self.save_data()
+
+# Configure logging
```
Это лучше выносить в utils.py или utils/logger.py и тп

TheMoosnter replied:
Не понял конкретно что выносить

TheMoosnter replied:
А, подожди, выносить конфигурацию логов? Типо настройку логов перенести в utils.py?

hnkovr replied:
да, коммент всегда пишется на строку (вообще иногда лушче на блок кода – посмотрю, как в гитхабе это можно...)

TheMoosnter replied:
Я опять сбился с суть проблемы. Так в чём проблема? Комментарий # Configure logging как-то не так написан, или сама настройка логгирования не в том файле прописана?

hnkovr replied:
а, я понял:
- это и есть config.py 😅
- тогда тут всё что ниже это строки – норм, а выше – надо куда-то в другой модуль вынести - это не конфиг-код
  - делай в pycharm через shift+F6 что ли...

hnkovr commented on config.py:
```diff
@@ -1,23 +1,119 @@
 CHAT_ID = os.getenv("CHAT_ID")
 
+# File for saving program variables
 DATA_FILE = "data.yaml"
```
А почему этот остался, если теперь студенты в csv?

TheMoosnter replied:
В этом файле хранятся значения переменных: какой индекс ученика, который сейчас дежурит, и список учеников, которые пропустили дежурство. То есть в этом файле хранятся переменные програмыы, а чтобы при выключении программы все эти данные не пропали, я их сохраняю в файл

hnkovr replied:
а, тут Ок

но в будущем такое надо всегда в БД хранить

hnkovr commented on utils.py:
```diff
@@ -108,12 +88,12 @@ def process_set(self, student_name):
         :param student_name: Student name.
         """
-        if student_name in STUDENT_LIST:
+        if student_name in self.student_list:
```
На подумать:
- код сильно похож на process_put
- нельзя ли выделить общий базовый метод?

Это про принцип DRY - погугли

TheMoosnter replied:
Можно создать функцию, которая будет проверять, есть ли передаваемый параметр (ученик) в списке, который тоже передаётся параметром (либо весь список учеников (в случае команды /set), либо список пропустивших дежурство учеников (/put))

hnkovr replied:
ещё варик, когда думать неохота:
- закидывать подобный код в ЛЛМ
- и писать что-то типа "refact this code (maybe create some common method to reduce code repeating)
;)

hnkovr commented on config.py:
```diff
@@ -1,23 +1,119 @@
-    "Sophia",
-]
+# Loading of student list from .cvs file to list constant
+STUDENT_LIST = load_students("students.csv")
```
Скаляр захардкожен... 

Ты ещё не пробовал использовать config.yaml? 

Ну или хотя бы в config.py его перенеси. 


Code style тоже хромает: 
- посмотри в PEP-8 для чего капс используется в питоне
- ruff не прогонял ещё? По идее он может такое обнаруживать.

TheMoosnter replied:
Не совсем понял. Это константа - список учеников, который можно изменить только вручную, редактируя файл .csv. Редактирование файла во время работы программы не предусмотрено, так что не совсем понял суть проблемы

hnkovr replied:
ну чуть более правильно тогда так:
DATA_FNAME = "students.csv"
STUDENT_LIST = load_students(DATA_FNAME)

ещё правильней:
```python
class Config:
  data_fname = "students.csv"
  student_list_data = load_students(data_fname)
``` 
или лучше вместо последнего, напр, что-то типа (но здесь уже больше вариаций возможных)
```python
class StudentList:  
  data = load_students(data_fname)
```     
ещё правильней:
config.yaml
```yaml
data:
  students_fname: students.csv
...
```   

config.py
```python
class Config:
   ...
   data_fname = config['data_fname']

```

TheMoosnter replied:
Типо сделать 
config.yaml
data:
data_fname: students.csv
для того, чтобы пользователь мог изменить название файла при желании? Можно, но я думаю это не особо функционально

hnkovr replied:
тут да, не нужно – пригодится в более серьёзных проектах 
- тч тут чисто учебный кейс

hnkovr replied:
> Не совсем понял. Это константа - список учеников, который можно изменить только вручную, редактируя файл .csv. Редактирование файла во время работы программы не предусмотрено, так что не совсем понял суть проблемы

внешние данные никогда не считают константой 
- тут это сложновато объянсить, но на будущее
- обычно константы - это какие-то скаляры, объявленные прямо в коде
- ну мб ещё какие-то настройки, загруженные из конфига

ПоГПТи попросив примеры

TheMoosnter committed:
new func for check + error message in PS

new function for checking exist of student in list + sending a message with error in private messages

hnkovr commented:
Code style тоже хромает:
- посмотри в PEP-8 для чего капс используется в питоне
- ruff не прогонял ещё? По идее он может такое обнаруживать.

\* отдельно это тоже добавлю, как общий коммент

TheMoosnter committed:
code format by black

TheMoosnter committed:
import sort by isort

hnkovr reviewed (commented):
дал комментов в открытых тредах
